close 1078 option cleanup (#1109)
exercism/rust: concepts/option/about.md

> The `Option` type is an enum provided by the standard library. It is useful for expressing whether a value exists or does not exist.

Reviewer: ```suggestion
The `Option` type is an enum provided by the standard library. It is useful for expressing a value which may not exist.
```

Let's be precise.

While this may sound like `null` or `nil`, it is quite different and special.

> `Option` is distinctive to Rust. It is also quite fun to work with!

Reviewer: > `Option` is distinctive to Rust.

Not true. For example, [`Maybe` in Haskell](https://wiki.haskell.org/Maybe). I'm pretty sure it exists in the ML family of languages also.

Might be better to phrase as something like "`Option` builds on language design from functional languages." Or maybe just omit that sentence.

Author: Good to know; Thanks! I went with omitting for now.

> `Option<T>` holds other types and has two variants `Some` and `None`.

Reviewer: ```suggestion
`Option<T>` holds other types and has two variants: `Some(T)` and `None`.
```

just to be a little more precise.

For example, `Option<u32>` holds the `u32` type.

> The below snippets demonstrate what this would look like in some Rust code.

Reviewer: Suggest rewriting this line and below as 

```markdown
## Usage

### Using the `T` within `Option<T>`

There are several common idioms for using the value which might be present in an `Option`:

> TODO: write a short explanatory paragraph about each of these

- `if let Some(t) = optional { ... }`. 
- `match optional { Some(t) => {...}, None => {...}}`
- `optional.map(|t| ...)` **Note**: operates on the contained value if present. The binding becomes an `Option<U>`, where the given closure is `FnOnce(T) -> U`.
- `optional.unwrap()`. **Important**: Unwrapping can ease development in the short term or in test code, but it introduces the possibility of runtime panics. It should be avoided when possible.

## Examples

// the examples already present
```

Author: Great, thanks! I am not familiar with using `optional.map(|t| ...)` so will read up on that later today to discuss usage.

The `Option` type is an enum provided by the standard library. It is useful for expressing a value which may not exist.
While this may sound like `null` or `nil`, it is quite different and special.

`Option<T>` holds other types and has two variants: `Some(T)` and `None`.

For example, `Option<u32>` holds the `u32` type.

## Usage

### Using the `T` within `Option<T>`

There are several common idioms for using the value which might be present in an `Option`:

- `if let Some(t) = optional { ... }`. 

    The `if let` idiom extracts the value from an Option<T> and binds it to the variable `t`. This lets you use it directly in that block scope.

- `match optional { Some(t) => {...}, None => {...}}`
   
    The `match` operator also lets you unpack and handle `None` cases in one statement or expression. It is an alternative to a chain of if and else statements. Using `match` can increase the readability of some sections of code.
- `optional.map(|t| ...)` 

>    `Option::map` creates an Option with a new type created from the provided function. 

Reviewer: ```suggestion
   `Option::map` updates an Option in place, potentially with a new type, from the provided function. 
```

You don't have to change the type, but you can. 

```rust
assert_eq!(Some(2).map(|n| n * 2), Some(4));
assert_eq!(Some(2).map(|n| n.to_string()), Some("2".into()));
```
